File: src/qm/banking/domain/singlepayments/implementation/SinglePaymentsImplementation.java
package qm.banking.domain.singlepayments.implementation;

import qm.banking.domain.entities.Account;
import qm.banking.domain.singlepayments.api.Balance;
import qm.banking.domain.singlepayments.api.ExternalAccount;
import qm.banking.domain.singlepayments.api.InternalAccount;
import qm.banking.domain.singlepayments.api.SinglePayments;
import qm.banking.domain.singlepayments.api.AccountAccess;

public class SinglePaymentsImplementation implements SinglePayments {

    private final AccountAccess accountAccess;

    public SinglePaymentsImplementation(AccountAccess accountAccess) {
        this.accountAccess = accountAccess;
    }

    @Override
    public void internalTransfer(InternalAccount from, InternalAccount to, int amount) {

    }

    @Override
    public void externalTransfer(InternalAccount from, ExternalAccount to, int amount) {

    }

    @Override
    public Balance currentBalance(InternalAccount accountDef) {
        Account account = accountAccess.retrieveAccount(accountDef);
        return BalanceBuilder.balance(account);
    }
}

File: src/main/qm/banking/domain/singlepayments/implementation/SinglePaymentsImplementation.kt
package qm.banking.domain.singlepayments.implementation

import qm.banking.domain.entities.LedgerEntry
import qm.banking.domain.singlepayments.api.*

class SinglePaymentsImplementation(private val accountAccess: AccountAccess) : SinglePayments {

    override fun internalTransfer(from: InternalAccount, to: InternalAccount, amount: Int) {
        val fromAccount = accountAccess.retrieveAccount(from)
        val toAccount = accountAccess.retrieveAccount(to)
        fromAccount.debit(amount)
        toAccount.credit(amount)
        val fromJournal = accountAccess.retrieveJournal(from)
        val toJournal = accountAccess.retrieveJournal(to)
        fromJournal.debit(amount, "")
        toJournal.credit(amount, "")
        val ledger = accountAccess.retrieveLedger()
        ledger.addEntry(LedgerEntry(amount, from, to))
    }

    override fun externalTransfer(from: InternalAccount, to: ExternalAccount, amount: Int) {
        throw IllegalStateException("not implemented")
    }

    override fun currentBalance(account: InternalAccount): Balance {
        val actualAccount = accountAccess.retrieveAccount(account)
        return BalanceBuilder.balance(actualAccount)
    }

}